{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run ../../bin/imports.ipynb\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gather pre-post immunotherapy neoantigen data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*** This won't work until patient data is moved to new location ***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Helper methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pipeline_file_other(x):\n",
    "    neo_home = x[0]\n",
    "    sample = x[1]\n",
    "    try:\n",
    "        reporting_dir = neo_home.split('/Pipeline_Runs')[0] + '/Reports/'\n",
    "        return reporting_dir + [x for x in os.listdir(reporting_dir) if (sample.split('_')[0] in x) and \\\n",
    "                                            (sample.split('_')[1].lower() in x.lower())][0] + '/DNA_Pipeline/Annotated_SmallVariant_Reports/All_Transcripts/{0}_somatic_dna_small_variant_report_all.xlsx'.format(sample)\n",
    "    except:\n",
    "        return '-'\n",
    "\n",
    "def get_stripped_sample(x):\n",
    "    if 'DNA' in x[:4]:\n",
    "        return x[4:]\n",
    "    else:\n",
    "        return x[:-4]\n",
    "    \n",
    "def get_hla_file(x):\n",
    "    QC_dir = x + '/QC'\n",
    "    return QC_dir + '/' + [x for x in os.listdir(QC_dir) if 'HLA_type.txt' in x][0]\n",
    "\n",
    "def get_refseq(x):\n",
    "    path1 = '{0}/Transcript_Estimation/RNA_{1}/Refseq/quant.sf'.format(x[0], x[1])\n",
    "    path2 = '{0}/Transcript_Estimation/{1}_RNA/Refseq/quant.sf'.format(x[0], x[1])\n",
    "    path3 = '{0}/Transcript_Estimation/{1}/Refseq/quant.sf'.format(x[0], x[1])\n",
    "    path4 = '{0}/Transcript_Estimation/{1}RNA/Refseq/quant.sf'.format(x[0], x[1][:-4])\n",
    "    if os.path.exists(path1):\n",
    "        return path1\n",
    "    elif os.path.exists(path2):\n",
    "        return path2\n",
    "    elif os.path.exists(path3):\n",
    "        return path3\n",
    "    else:\n",
    "        return path4\n",
    "    \n",
    "def get_ensembl(x):\n",
    "    path1 = '{0}/Transcript_Estimation/RNA_{1}/Ensembl/quant.sf'.format(x[0], x[1])\n",
    "    path2 = '{0}/Transcript_Estimation/{1}_RNA/Ensembl/quant.sf'.format(x[0], x[1])\n",
    "    path3 = '{0}/Transcript_Estimation/{1}/Ensembl/quant.sf'.format(x[0], x[1])\n",
    "    path4 = '{0}/Transcript_Estimation/{1}RNA/Ensembl/quant.sf'.format(x[0], x[1][:-4])\n",
    "    if os.path.exists(path1):\n",
    "        return path1\n",
    "    elif os.path.exists(path2):\n",
    "        return path2\n",
    "    elif os.path.exists(path3):\n",
    "        return path3\n",
    "    else:\n",
    "        return path4\n",
    "    \n",
    "def get_fusion_catcher(x):\n",
    "    path = '{0}/FusionCatcher/RNA_{1}/final-list_candidate-fusion-genes.GRCh37.txt'.format(x[0], x[1])\n",
    "    if os.path.exists(path):\n",
    "        return path\n",
    "    else:\n",
    "        return 'Fail'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Gather data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_df = pd.read_csv('/rnd/users/rpyke/data/00-DASH/data/readdata.input_df.nextDB_collaboration_2020_09_11.csv',\n",
    "           index_col=0)\n",
    "\n",
    "input_df = input_df[input_df.group.isin(['stanford_sunwoo'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cancer_home</th>\n",
       "      <th>group</th>\n",
       "      <th>neo_home</th>\n",
       "      <th>normal_home</th>\n",
       "      <th>rna_home</th>\n",
       "      <th>sample_name_normal</th>\n",
       "      <th>sample_name_tumor</th>\n",
       "      <th>tcr_home</th>\n",
       "      <th>tumor_home</th>\n",
       "      <th>hla_type</th>\n",
       "      <th>sequenza_confints_cp</th>\n",
       "      <th>hla_somatic_mutations</th>\n",
       "      <th>normal_qc_summary</th>\n",
       "      <th>tumor_qc_summary</th>\n",
       "      <th>sequenza_cnv</th>\n",
       "      <th>normal_fastq</th>\n",
       "      <th>tumor_fastq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>stanford_sunwoo</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>40425-0012_Post_PBMC_DNA</td>\n",
       "      <td>40425-0012_Pre_Tumor_DNA</td>\n",
       "      <td>-</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>stanford_sunwoo</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>40425-0008_Pre_PBMC_DNA</td>\n",
       "      <td>40425-0008_Pre_Tumor_DNA</td>\n",
       "      <td>-</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>stanford_sunwoo</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>40425-0014_Pre_PBMC_DNA</td>\n",
       "      <td>40425-0014_Post_Tumor_DNA</td>\n",
       "      <td>-</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>stanford_sunwoo</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>40425-0014_Pre_PBMC_DNA</td>\n",
       "      <td>40425-0014_Pre_Tumor_DNA</td>\n",
       "      <td>-</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>stanford_sunwoo</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>40425-0001_Post_PBMC_DNA</td>\n",
       "      <td>40425-0001_Post_Tumor_DNA</td>\n",
       "      <td>-</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "      <td>/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          cancer_home            group  \\\n",
       "10  /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...  stanford_sunwoo   \n",
       "8   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...  stanford_sunwoo   \n",
       "7   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...  stanford_sunwoo   \n",
       "6   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...  stanford_sunwoo   \n",
       "5   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...  stanford_sunwoo   \n",
       "\n",
       "                                             neo_home  \\\n",
       "10  /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "8   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "7   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "6   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "5   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "\n",
       "                                          normal_home  \\\n",
       "10  /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "8   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "7   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "6   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "5   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "\n",
       "                                             rna_home  \\\n",
       "10  /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "8   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "7   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "6   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "5   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "\n",
       "          sample_name_normal          sample_name_tumor tcr_home  \\\n",
       "10  40425-0012_Post_PBMC_DNA   40425-0012_Pre_Tumor_DNA        -   \n",
       "8    40425-0008_Pre_PBMC_DNA   40425-0008_Pre_Tumor_DNA        -   \n",
       "7    40425-0014_Pre_PBMC_DNA  40425-0014_Post_Tumor_DNA        -   \n",
       "6    40425-0014_Pre_PBMC_DNA   40425-0014_Pre_Tumor_DNA        -   \n",
       "5   40425-0001_Post_PBMC_DNA  40425-0001_Post_Tumor_DNA        -   \n",
       "\n",
       "                                           tumor_home  \\\n",
       "10  /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "8   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "7   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "6   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "5   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "\n",
       "                                             hla_type  \\\n",
       "10  /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "8   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "7   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "6   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "5   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "\n",
       "                                 sequenza_confints_cp  \\\n",
       "10  /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "8   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "7   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "6   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "5   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "\n",
       "                                hla_somatic_mutations  \\\n",
       "10  /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "8   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "7   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "6   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "5   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "\n",
       "                                    normal_qc_summary  \\\n",
       "10  /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "8   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "7   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "6   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "5   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "\n",
       "                                     tumor_qc_summary  \\\n",
       "10  /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "8   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "7   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "6   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "5   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "\n",
       "                                         sequenza_cnv  \\\n",
       "10  /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "8   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "7   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "6   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "5   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "\n",
       "                                         normal_fastq  \\\n",
       "10  /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "8   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "7   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "6   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "5   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...   \n",
       "\n",
       "                                          tumor_fastq  \n",
       "10  /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...  \n",
       "8   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...  \n",
       "7   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...  \n",
       "6   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...  \n",
       "5   /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Ru...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0012/Post_PBMC_DNA_Core_20181004/QC'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-36a0b6aaa62e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0minput_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'dna_pipeline'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minput_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'neo_home'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'sample_name_tumor'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mget_pipeline_file_other\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0minput_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'sample_stripped'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minput_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msample_name_tumor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mget_stripped_sample\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m#input_df.sample_name_tumor.str[4:]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0minput_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'HLA_types_file'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minput_df\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'normal_home'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mget_hla_file\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;31m# Get refseq transcript $rna_home/Transcript_Estimation/RNA_$sample_stripped/Refseq/quant.sf\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda2/envs/py37/lib/python3.7/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, func, convert_dtype, args, **kwds)\u001b[0m\n\u001b[1;32m   3589\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3590\u001b[0m                 \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3591\u001b[0;31m                 \u001b[0mmapped\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap_infer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconvert\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mconvert_dtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3592\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3593\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmapped\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmapped\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSeries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/lib.pyx\u001b[0m in \u001b[0;36mpandas._libs.lib.map_infer\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m<ipython-input-2-bf4e786a7452>\u001b[0m in \u001b[0;36mget_hla_file\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mget_hla_file\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     18\u001b[0m     \u001b[0mQC_dir\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mx\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'/QC'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 19\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mQC_dir\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'/'\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mx\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlistdir\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mQC_dir\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0;34m'HLA_type.txt'\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     20\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mget_refseq\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0012/Post_PBMC_DNA_Core_20181004/QC'"
     ]
    }
   ],
   "source": [
    "reporting_dir = list(input_df.neo_home)[0].split('/Pipeline_Runs')[0] + '/Reports'\n",
    "\n",
    "input_df['dna_pipeline'] = input_df[['neo_home', 'sample_name_tumor']].apply(get_pipeline_file_other, axis=1)\n",
    "input_df['sample_stripped'] = input_df.sample_name_tumor.apply(get_stripped_sample) #input_df.sample_name_tumor.str[4:]\n",
    "input_df['HLA_types_file'] = input_df['normal_home'].apply(get_hla_file)\n",
    "\n",
    "# Get refseq transcript $rna_home/Transcript_Estimation/RNA_$sample_stripped/Refseq/quant.sf\n",
    "input_df['refseq_transcripts'] = input_df[['rna_home', 'sample_stripped']].apply(get_refseq, axis=1)\n",
    "\n",
    "# Get ensembl transcript (and check that it's there)\n",
    "input_df['ensembl_transcripts'] = input_df[['rna_home', 'sample_stripped']].apply(get_ensembl, axis=1)\n",
    "\n",
    "# Get Fusion catcher\n",
    "input_df['fusion_catcher_ffn'] = input_df[['rna_home', 'sample_stripped']].apply(get_fusion_catcher, axis=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_df.to_csv('/rnd/users/rpyke/data/03-public_MS/applications/input_df.Stanford_sunwoo.2020_09_11.tab',\n",
    "                sep='\\t', index=None)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check for existence of needed files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_df.to_csv('/rnd/users/rpyke/data/03-public_MS/applications/input_df.Stanford_sunwoo.tab',\n",
    "                sep='\\t', index=None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0012/Pre_Tumor_Neoantigen/Antigen_ID/tumor_peptides.details\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0008/Pre_Tumor_Neoantigen/Antigen_ID/tumor_peptides.details\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0014/Post_Tumor_Neoantigen/Antigen_ID/tumor_peptides.details\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0014/Pre_Tumor_Neoantigen/Antigen_ID/tumor_peptides.details\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0001/Post_Tumor_Neoantigen/Antigen_ID/tumor_peptides.details\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0001/Pre_Tumor_Neoantigen/Antigen_ID/tumor_peptides.details\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0002/Post_Tumor_Neoantigen/Antigen_ID/tumor_peptides.details\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0011/Pre_Tumor_Neoantigen/Antigen_ID/tumor_peptides.details\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0009/Post_Tumor_Neoantigen/Antigen_ID/tumor_peptides.details\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0009/Pre_Tumor_Neoantigen/Antigen_ID/tumor_peptides.details\n",
      "-/Antigen_ID/tumor_peptides.details\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/7140/Post_Tumor_Neoantigen/Antigen_ID/tumor_peptides.details\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0013/Pre_Tumor_Neoantigen/Antigen_ID/tumor_peptides.details\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0013/Post_Tumor_Neoantigen/Antigen_ID/tumor_peptides.details\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0008/Post_Tumor_Neoantigen/Antigen_ID/tumor_peptides.details\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0002/Pre_Tumor_Neoantigen/Antigen_ID/tumor_peptides.details\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0012/Pre_Tumor_Neoantigen/Antigen_ID/SNVIndel/tumor_peptides_protein.fa\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0008/Pre_Tumor_Neoantigen/Antigen_ID/SNVIndel/tumor_peptides_protein.fa\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0014/Post_Tumor_Neoantigen/Antigen_ID/SNVIndel/tumor_peptides_protein.fa\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0014/Pre_Tumor_Neoantigen/Antigen_ID/SNVIndel/tumor_peptides_protein.fa\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0001/Post_Tumor_Neoantigen/Antigen_ID/SNVIndel/tumor_peptides_protein.fa\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0001/Pre_Tumor_Neoantigen/Antigen_ID/SNVIndel/tumor_peptides_protein.fa\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0002/Post_Tumor_Neoantigen/Antigen_ID/SNVIndel/tumor_peptides_protein.fa\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0011/Pre_Tumor_Neoantigen/Antigen_ID/SNVIndel/tumor_peptides_protein.fa\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0009/Post_Tumor_Neoantigen/Antigen_ID/SNVIndel/tumor_peptides_protein.fa\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0009/Pre_Tumor_Neoantigen/Antigen_ID/SNVIndel/tumor_peptides_protein.fa\n",
      "-/Antigen_ID/SNVIndel/tumor_peptides_protein.fa\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/7140/Post_Tumor_Neoantigen/Antigen_ID/SNVIndel/tumor_peptides_protein.fa\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0013/Pre_Tumor_Neoantigen/Antigen_ID/SNVIndel/tumor_peptides_protein.fa\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0013/Post_Tumor_Neoantigen/Antigen_ID/SNVIndel/tumor_peptides_protein.fa\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0008/Post_Tumor_Neoantigen/Antigen_ID/SNVIndel/tumor_peptides_protein.fa\n",
      "/rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0002/Pre_Tumor_Neoantigen/Antigen_ID/SNVIndel/tumor_peptides_protein.fa\n",
      "40425-0012_Pre_Tumor /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0012/Pre_Tumor_RNA_CancerRNA_20181014/Transcript_Estimation/RNA_40425-0012_Pre_Tumor/Ensembl/quant.sf\n",
      "40425-0008_Pre_Tumor /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0008/Pre_Tumor_RNA_CancerRNA_20181014/Transcript_Estimation/RNA_40425-0008_Pre_Tumor/Ensembl/quant.sf\n",
      "40425-0014_Post_Tumor /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0014/Post_Tumor_RNA_CancerRNA_20181022/Transcript_Estimation/RNA_40425-0014_Post_Tumor/Ensembl/quant.sf\n",
      "40425-0014_Pre_Tumor /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0014/Pre_Tumor_RNA_CancerRNA_20181014/Transcript_Estimation/RNA_40425-0014_Pre_Tumor/Ensembl/quant.sf\n",
      "40425-0001_Post_Tumor /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0001/Post_Tumor_RNA_CancerRNA_20181014/Transcript_Estimation/RNA_40425-0001_Post_Tumor/Ensembl/quant.sf\n",
      "40425-0001_Pre_TumorDNA_ /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0001/Pre_Tumor_RNA_CancerRNA_20181014/Transcript_Estimation/RNA_40425-0001_Pre_TumorDNA_/Ensembl/quant.sf\n",
      "40425-0002_Post_Tumor /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0002/Post_Tumor_RNA_CancerRNA_20181014/Transcript_Estimation/RNA_40425-0002_Post_Tumor/Ensembl/quant.sf\n",
      "40425-0011_Pre_Tumor /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0011/Pre_Tumor_RNA_CancerRNA_20181014/Transcript_Estimation/RNA_40425-0011_Pre_Tumor/Ensembl/quant.sf\n",
      "40425-0009_Post_Tumor /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0009/Post_Tumor_RNA_CancerRNA_20181014/Transcript_Estimation/RNA_40425-0009_Post_Tumor/Ensembl/quant.sf\n",
      "40425-0009_Pre_Tumor /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0009/Pre_Tumor_RNA_CancerRNA_20181014/Transcript_Estimation/RNA_40425-0009_Pre_Tumor/Ensembl/quant.sf\n",
      "40425-0011_Post_Tumor -/Transcript_Estimation/RNA_40425-0011_Post_Tumor/Ensembl/quant.sf\n",
      "7140_Post_Tumor /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/7140/Post_Tumor_RNA_CancerRNA_20181014/Transcript_Estimation/RNA_7140_Post_Tumor/Ensembl/quant.sf\n",
      "40425-0013_Pre_Tumor /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0013/Pre_Tumor_RNA_CancerRNA_20181014/Transcript_Estimation/RNA_40425-0013_Pre_Tumor/Ensembl/quant.sf\n",
      "40425-0013_Post_Tumor /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0013/Post_Tumor_RNA_CancerRNA_20181014/Transcript_Estimation/RNA_40425-0013_Post_Tumor/Ensembl/quant.sf\n",
      "40425-0008_Post_Tumor /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0008/Post_Tumor_RNA_CancerRNA_20181014/Transcript_Estimation/RNA_40425-0008_Post_Tumor/Ensembl/quant.sf\n",
      "40425-0002_Pre_TumorDNA_ /rnd/Collaboration/Stanford_Sunwoo/Pipeline_Runs/40425-0002/Pre_Tumor_RNA_CancerRNA_20181014/Transcript_Estimation/RNA_40425-0002_Pre_TumorDNA_/Ensembl/quant.sf\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'HLA_types_file'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-18-7c3faafa2edd>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[0;31m# $hla_type\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 25\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mHLA_types_file\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mHLA_types_file\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     26\u001b[0m     \u001b[0mpath\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mHLA_types_file\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexists\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda2/envs/py37/lib/python3.7/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m__getattr__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   5065\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5066\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5067\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5068\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5069\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__setattr__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'HLA_types_file'"
     ]
    }
   ],
   "source": [
    "# $neo_home/Antigen_ID/tumor_peptides.details\n",
    "for neo_home in list(input_df.neo_home):\n",
    "    path = '{0}/Antigen_ID/tumor_peptides.details'.format(neo_home)\n",
    "    if not os.path.exists(path):\n",
    "        print(path)\n",
    "\n",
    "# $neo_home/Antigen_ID/SNVIndel/tumor_peptides_protein.fa\n",
    "for neo_home in list(input_df.neo_home):\n",
    "    path = '{0}/Antigen_ID/SNVIndel/tumor_peptides_protein.fa'.format(neo_home)\n",
    "    if not os.path.exists(path):\n",
    "        print(path)\n",
    "\n",
    "# $rna_home/Transcript_Estimation/RNA_$sample_stripped/Ensembl/quant.sf\n",
    "for rna_home, sample_stripped in zip(list(input_df.rna_home), list(input_df.sample_stripped)):\n",
    "    path1 = '{0}/Transcript_Estimation/RNA_{1}/Ensembl/quant.sf'.format(rna_home, sample_stripped)\n",
    "    path2 = '{0}/Transcript_Estimation/{1}_RNA/Ensembl/quant.sf'.format(rna_home, sample_stripped)\n",
    "    path3 = '{0}/Transcript_Estimation/{1}/Ensembl/quant.sf'.format(rna_home, sample_stripped)\n",
    "    path4 = '{0}/Transcript_Estimation/{1}RNA/Ensembl/quant.sf'.format(rna_home, sample_stripped[:-4])\n",
    " \n",
    "    if not(os.path.exists(path1) or os.path.exists(path2) or os.path.exists(path3) or os.path.exists(path4)):\n",
    "        print(sample_stripped, path1)\n",
    "        \n",
    "\n",
    "# $hla_type\n",
    "for HLA_types_file in list(input_df.HLA_types_file):\n",
    "    path = HLA_types_file\n",
    "    if not os.path.exists(path):\n",
    "        print(path)\n",
    "\n",
    "# $dna_pipeline\n",
    "for dna_pipeline in list(input_df.dna_pipeline):\n",
    "    path = dna_pipeline\n",
    "    if not os.path.exists(path):\n",
    "        print(path)\n",
    "\n",
    "# $neo_home/Known_Antigens/all_kmer_candidates.fa \n",
    "for neo_home in list(input_df.neo_home):\n",
    "    path = '{0}/Known_Antigens/all_kmer_candidates.fa'.format(neo_home)\n",
    "    if not os.path.exists(path):\n",
    "        print(path)\n",
    "\n",
    "# $dna_pipeline\n",
    "for refseq_transcript in list(input_df.refseq_transcripts):\n",
    "    path = refseq_transcript\n",
    "    if not os.path.exists(path):\n",
    "        print(path)\n",
    "\n",
    "# $dna_pipeline\n",
    "for ensembl_transcript in list(input_df.ensembl_transcripts):\n",
    "    path = ensembl_transcript\n",
    "    if not os.path.exists(path):\n",
    "        print(path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(input_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_cluster_script(samples, normal_homes, rna_homes, neo_homes, dna_pipelines, samples_stripped, \n",
    "                          hla_types, ensembl, refseq, fusion_catcher):\n",
    "\n",
    "    new_script_file = '/rnd/users/rpyke/code/03-public_MS/scripts/prepare_sherpa.sh'\n",
    "    \n",
    "    with open(new_script_file, 'w') as out_file:\n",
    "        out_file.write(\"#! /bin/csh\\n\")\n",
    "        out_file.write(\"#$ -S /bin/csh\\n\")\n",
    "        out_file.write(\"#$ -cwd\\n\")\n",
    "        out_file.write(\"#$ -V\\n\")\n",
    "        out_file.write(\"#$ -q all.q\\n\")\n",
    "        out_file.write(\"#$ -pe shm 1\\n\")\n",
    "        out_file.write(\"#$ -l h_vmem=30G\\n\")\n",
    "        out_file.write(\"#$ -N SHERPA\\n\")\n",
    "        #out_file.write(\"#$ -t 1-2\\n\".format(len(samples)))\n",
    "        out_file.write(\"#$ -t 1-{0}\\n\".format(len(samples)))\n",
    "        out_file.write(\"#$ -tc 2\\n\")\n",
    "        out_file.write(\"#$ -o /rnd/users/rpyke/data/03-public_MS/sge_system_files\\n\")\n",
    "        out_file.write(\"#$ -e /rnd/users/rpyke/data/03-public_MS/sge_system_files\\n\")\n",
    "        out_file.write(\"\\n\")\n",
    "\n",
    "        out_file.write(\"set samples=({0})\\n\".format(\" \".join(samples)))\n",
    "        out_file.write(\"set normal_homes=({0})\\n\".format(\" \".join(normal_homes)))\n",
    "        out_file.write(\"set rna_homes=({0})\\n\".format(\" \".join(rna_homes)))\n",
    "        out_file.write(\"set neo_homes=({0})\\n\".format(\" \".join(neo_homes)))\n",
    "        out_file.write(\"set dna_pipelines=({0})\\n\".format(\" \".join(dna_pipelines)))\n",
    "        out_file.write(\"set samples_stripped=({0})\\n\".format(\" \".join(samples_stripped)))\n",
    "        out_file.write(\"set hla_types=({0})\\n\".format(\" \".join(hla_types)))\n",
    "        out_file.write(\"set ensembls=({0})\\n\".format(\" \".join(ensembl)))\n",
    "        out_file.write(\"set refseqs=({0})\\n\".format(\" \".join(refseq)))\n",
    "        out_file.write(\"set fusions=({0})\\n\".format(\" \".join(fusion_catcher)))\n",
    "        out_file.write(\"\\n\")\n",
    "\n",
    "        out_file.write(\"set sample=$samples[$SGE_TASK_ID]\\n\")\n",
    "        out_file.write(\"set normal_home=$normal_homes[$SGE_TASK_ID]\\n\")\n",
    "        out_file.write(\"set rna_home=$rna_homes[$SGE_TASK_ID]\\n\")\n",
    "        out_file.write(\"set neo_home=$neo_homes[$SGE_TASK_ID]\\n\")\n",
    "        out_file.write(\"set dna_pipeline=$dna_pipelines[$SGE_TASK_ID]\\n\")\n",
    "        out_file.write(\"set sample_stripped=$samples_stripped[$SGE_TASK_ID]\\n\")\n",
    "        out_file.write(\"set hla_type=$hla_types[$SGE_TASK_ID]\\n\")\n",
    "        out_file.write(\"set ensembl=$ensembls[$SGE_TASK_ID]\\n\")\n",
    "        out_file.write(\"set refseq=$refseqs[$SGE_TASK_ID]\\n\")\n",
    "        out_file.write(\"set fusion=$fusions[$SGE_TASK_ID]\\n\")\n",
    "        out_file.write(\"\\n\")\n",
    "\n",
    "        out_file.write(\"date\\n\")\n",
    "        out_file.write(\"hostname\\n\")\n",
    "        out_file.write(\"\\n\")\n",
    "        \n",
    "        #out_file.write(\"mkdir /rnd/users/rpyke/data/03-public_MS/inova/sherpa_input/$sample\\n\")\n",
    "        \n",
    "        out_file.write(\"echo python /rnd/users/mp-wsrv-hmc01/sherpa/sherpa/gather_sherpa_features.py \" + \\\n",
    "                        \"--snv_fusion_details $neo_home/Antigen_ID/tumor_peptides.details \" + \\\n",
    "                        \"--snvindel_tumor_protein_fasta $neo_home/Antigen_ID/SNVIndel/tumor_peptides_protein.fa \" + \\\n",
    "                        #\"--tumor_super_peptide_fasta $neo_home/Antigen_ID/tumor_peptides.fa \" + \\\n",
    "                        \"--ensembl_transcript_exp $ensembl \" + \\\n",
    "                        \"--refseq_transcript_exp $refseq \" +\\\n",
    "                        \"--bpseq_ffn /rnd/users/mp-wsrv-hmc01/sherpa/data/bpseq.tab \" + \\\n",
    "                        \"--hla_alleles $hla_type \" + \\\n",
    "                        \"--snvindel_all_transcript_ffn $dna_pipeline \" + \\\n",
    "                        \"--gene_prop_ffn /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.1.0/gene_prop.tab \" + \\\n",
    "                        \"--hotspot_ffn /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.1.0/hotspot.tab \" + \\\n",
    "                        \"--all_kmer_candidates $neo_home/Known_Antigens/all_kmer_candidates.fa \" + \\\n",
    "                        \"--fusion_catcher_ffn $fusion \" + \\\n",
    "                        \"--fasta_out /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/temp_fastas/$sample.fa \" + \\\n",
    "                        \"--sherpa_prediction_input /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_input/$sample.tab\\n\")\n",
    "        out_file.write(\"python /rnd/users/mp-wsrv-hmc01/sherpa/sherpa/gather_sherpa_features.py \" + \\\n",
    "                        \"--snv_fusion_details $neo_home/Antigen_ID/tumor_peptides.details \" + \\\n",
    "                        \"--snvindel_tumor_protein_fasta $neo_home/Antigen_ID/SNVIndel/tumor_peptides_protein.fa \" + \\\n",
    "                        #\"--tumor_super_peptide_fasta $neo_home/Antigen_ID/tumor_peptides.fa \" + \\\n",
    "                        \"--ensembl_transcript_exp $ensembl \" + \\\n",
    "                        \"--refseq_transcript_exp $refseq \" +\\\n",
    "                        \"--bpseq_ffn /rnd/users/mp-wsrv-hmc01/sherpa/data/bpseq.tab \" + \\\n",
    "                        \"--hla_alleles $hla_type \" + \\\n",
    "                        \"--snvindel_all_transcript_ffn $dna_pipeline \" + \\\n",
    "                        \"--gene_prop_ffn /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.1.0/gene_prop.tab \" + \\\n",
    "                        \"--hotspot_ffn /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.1.0/hotspot.tab \" + \\\n",
    "                        \"--all_kmer_candidates $neo_home/Known_Antigens/all_kmer_candidates.fa \" + \\\n",
    "                        \"--fusion_catcher_ffn $fusion \" + \\\n",
    "                        \"--fasta_out /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/temp_fastas/$sample.fa \" + \\\n",
    "                        \"--sherpa_prediction_input /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_input/$sample.tab\\n\")\n",
    "        out_file.write(\"echo Sherpa data prepared.\\n\")\n",
    "        \n",
    "                              \n",
    "        out_file.write(\"date\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "create_cluster_script(list(input_df.sample_name_tumor),\n",
    "                     list(input_df.normal_home),\n",
    "                     list(input_df.rna_home),\n",
    "                     list(input_df.neo_home),\n",
    "                     list(input_df.dna_pipeline),\n",
    "                     list(input_df.sample_stripped),\n",
    "                     list(input_df.HLA_types_file),\n",
    "                     list(input_df.ensembl_transcripts),\n",
    "                     list(input_df.refseq_transcripts),\n",
    "                     list(input_df.fusion_catcher_ffn))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Running the scripts through the sherpa model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Presentation model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_cluster_script_negative(patients):\n",
    "    \n",
    "    new_script_file = '/rnd/users/rpyke/code/03-public_MS/scripts/run_sherpa.nextdb.presentation.sh'\n",
    "    \n",
    "    with open(new_script_file, 'w') as out_file:\n",
    "        out_file.write(\"#! /bin/csh\\n\")\n",
    "        out_file.write(\"#$ -S /bin/csh\\n\")\n",
    "        out_file.write(\"#$ -cwd\\n\")\n",
    "        out_file.write(\"#$ -V\\n\")\n",
    "        out_file.write(\"#$ -pe shm 4\\n\")\n",
    "        out_file.write(\"#$ -l h_vmem=10G\\n\")\n",
    "        out_file.write(\"#$ -N SHERPA_inova\\n\")\n",
    "        #out_file.write(\"#$ -t 1-1\\n\".format(len(projects)))\n",
    "        out_file.write(\"#$ -t 1-{0}\\n\".format(len(patients)))\n",
    "        out_file.write(\"#$ -tc 3\\n\")\n",
    "        out_file.write(\"#$ -q all.q\\n\")\n",
    "        out_file.write(\"#$ -o /rnd/users/rpyke/data/03-public_MS/sge_system_files/\\n\")\n",
    "        out_file.write(\"#$ -e /rnd/users/rpyke/data/03-public_MS/sge_system_files\\n\")\n",
    "        out_file.write(\"\\n\")\n",
    "\n",
    "        out_file.write(\"set patients=({0})\\n\".format(\" \".join(patients)))\n",
    "        out_file.write(\"\\n\")\n",
    "    \n",
    "        out_file.write(\"set patient=$patients[$SGE_TASK_ID]\\n\")\n",
    "        out_file.write(\"\\n\")\n",
    "\n",
    "        out_file.write(\"date\\n\")\n",
    "        out_file.write(\"hostname\\n\")\n",
    "        out_file.write(\"echo $patient\")\n",
    "        out_file.write(\"\\n\")\n",
    "    \n",
    "        # SNV\n",
    "        out_file.write(\"cd /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/SNV/\\n\")\n",
    "        out_file.write(\"python /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/SNV/1595312117657302__20200720-222838_snv_production_inhouse__3383dd3a8357423e977d9b6327b9b6f8.predict.py \" + \\\n",
    "                        \"--in_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_input/$patient.snv.tab \" + \\\n",
    "                        \"--out_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_output/$patient.snv.PRESENTATION.v1.2.3.tab \" + \\\n",
    "                        \"--max_mem 30000000000 --max_cpus 4 --dataframe_backend parallel_pandas \" + \\\n",
    "                        \"--calibration_file /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/all_data.SNV.tab\\n\")   \n",
    "        out_file.write(\"\\n\")\n",
    "                  \n",
    "        # INDEL\n",
    "        out_file.write(\"cd /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/INDEL/\\n\")\n",
    "        out_file.write(\"python /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/INDEL/15953116605114079__20200720-222908_indel_production_inhouse__b0524b15001b4251b33d08a02bebcac3.predict.py \" + \\\n",
    "                        \"--in_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_input/$patient.indel.tab \" + \\\n",
    "                        \"--out_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_output/$patient.indel.PRESENTATION.v1.2.3.tab \" + \\\n",
    "                        \"--max_mem 30000000000 --max_cpus 4 --dataframe_backend parallel_pandas \" + \\\n",
    "                        \"--calibration_file /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/all_data.INDEL.tab\\n\")   \n",
    "        out_file.write(\"\\n\")\n",
    "                  \n",
    "        \n",
    "        # FUSION\n",
    "        out_file.write(\"cd /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/FUSION/\\n\")\n",
    "        out_file.write(\"python /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/FUSION/15953122502236447__20200720-223224_fusion_production_inhouse__e0351c9003b241eea2a197fe07603437.predict.py \" + \\\n",
    "                        \"--in_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_input/$patient.fusion.tab \" + \\\n",
    "                        \"--out_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_output/$patient.fusion.PRESENTATION.v1.2.3.tab \" + \\\n",
    "                        \"--max_mem 30000000000 --max_cpus 4 --dataframe_backend parallel_pandas \" + \\\n",
    "                        \"--calibration_file /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/all_data.FUSION.tab\\n\")      \n",
    "        out_file.write(\"\\n\")\n",
    "        \n",
    "        # BASELINE\n",
    "        out_file.write(\"cd /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/BINDING/\\n\")\n",
    "        out_file.write(\"python /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/BINDING/15953120345848932__20200720-222808_binding_production_inhouse__27c885651fd0497c876899abb3f4ea27.predict.py \" + \\\n",
    "                        \"--in_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_input/$patient.baseline.tab \" + \\\n",
    "                        \"--out_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_output/$patient.baseline.PRESENTATION.v1.2.3.tab \" + \\\n",
    "                        \"--max_mem 30000000000 --max_cpus 4 --dataframe_backend parallel_pandas \" + \\\n",
    "                        \"--calibration_file /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/all_data.BINDING.tab\\n\")   \n",
    "        out_file.write(\"\\n\")\n",
    "            \n",
    "        out_file.write(\"date\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "create_cluster_script_negative(list(input_df.sample_name_tumor))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "BINDING Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_cluster_script_negative(patients):\n",
    "    \n",
    "    new_script_file = '/rnd/users/rpyke/code/03-public_MS/scripts/run_sherpa.nextdb.binding.sh'\n",
    "    \n",
    "    with open(new_script_file, 'w') as out_file:\n",
    "        out_file.write(\"#! /bin/csh\\n\")\n",
    "        out_file.write(\"#$ -S /bin/csh\\n\")\n",
    "        out_file.write(\"#$ -cwd\\n\")\n",
    "        out_file.write(\"#$ -V\\n\")\n",
    "        out_file.write(\"#$ -pe shm 4\\n\")\n",
    "        out_file.write(\"#$ -l h_vmem=10G\\n\")\n",
    "        out_file.write(\"#$ -N SHERPA_inova\\n\")\n",
    "        #out_file.write(\"#$ -t 1-1\\n\".format(len(projects)))\n",
    "        out_file.write(\"#$ -t 1-{0}\\n\".format(len(patients)))\n",
    "        out_file.write(\"#$ -tc 3\\n\")\n",
    "        out_file.write(\"#$ -q all.q\\n\")\n",
    "        out_file.write(\"#$ -o /rnd/users/rpyke/data/03-public_MS/sge_system_files/\\n\")\n",
    "        out_file.write(\"#$ -e /rnd/users/rpyke/data/03-public_MS/sge_system_files\\n\")\n",
    "        out_file.write(\"\\n\")\n",
    "\n",
    "        out_file.write(\"set patients=({0})\\n\".format(\" \".join(patients)))\n",
    "        out_file.write(\"\\n\")\n",
    "    \n",
    "        out_file.write(\"set patient=$patients[$SGE_TASK_ID]\\n\")\n",
    "        out_file.write(\"\\n\")\n",
    "\n",
    "        out_file.write(\"date\\n\")\n",
    "        out_file.write(\"hostname\\n\")\n",
    "        out_file.write(\"echo $patient\")\n",
    "        out_file.write(\"\\n\")\n",
    "    \n",
    "        # SNV\n",
    "        out_file.write(\"cd /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/BINDING/\\n\")\n",
    "        out_file.write(\"python /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/BINDING/15953120345848932__20200720-222808_binding_production_inhouse__27c885651fd0497c876899abb3f4ea27.predict.py \" + \\\n",
    "                        \"--in_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_input/$patient.snv.tab \" + \\\n",
    "                        \"--out_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_output/$patient.snv.BINDING.v1.2.3.tab \" + \\\n",
    "                        \"--max_mem 30000000000 --max_cpus 4 --dataframe_backend parallel_pandas \" + \\\n",
    "                        \"--calibration_file /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/all_data.BINDING.tab\\n\")   \n",
    "        out_file.write(\"\\n\")\n",
    "                  \n",
    "        # INDEL\n",
    "        out_file.write(\"cd /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/BINDING/\\n\")\n",
    "        out_file.write(\"python /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/BINDING/15953120345848932__20200720-222808_binding_production_inhouse__27c885651fd0497c876899abb3f4ea27.predict.py \" + \\\n",
    "                        \"--in_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_input/$patient.indel.tab \" + \\\n",
    "                        \"--out_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_output/$patient.indel.BINDING.v1.2.3.tab \" + \\\n",
    "                        \"--max_mem 30000000000 --max_cpus 4 --dataframe_backend parallel_pandas \" + \\\n",
    "                        \"--calibration_file /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/all_data.BINDING.tab\\n\")   \n",
    "        out_file.write(\"\\n\")\n",
    "                  \n",
    "        \n",
    "        # FUSION\n",
    "        out_file.write(\"cd /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/BINDING/\\n\")\n",
    "        out_file.write(\"python /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/BINDING/15953120345848932__20200720-222808_binding_production_inhouse__27c885651fd0497c876899abb3f4ea27.predict.py \" + \\\n",
    "                        \"--in_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_input/$patient.fusion.tab \" + \\\n",
    "                        \"--out_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_output/$patient.fusion.BINDING.v1.2.3.tab \" + \\\n",
    "                        \"--max_mem 30000000000 --max_cpus 4 --dataframe_backend parallel_pandas \" + \\\n",
    "                        \"--calibration_file /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/all_data.BINDING.tab\\n\")      \n",
    "        out_file.write(\"\\n\")\n",
    "        \n",
    "        # BASELINE\n",
    "        out_file.write(\"cd /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/BINDING/\\n\")\n",
    "        out_file.write(\"python /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/BINDING/15953120345848932__20200720-222808_binding_production_inhouse__27c885651fd0497c876899abb3f4ea27.predict.py \" + \\\n",
    "                        \"--in_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_input/$patient.baseline.tab \" + \\\n",
    "                        \"--out_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_output/$patient.baseline.BINDING.v1.2.3.tab \" + \\\n",
    "                        \"--max_mem 30000000000 --max_cpus 4 --dataframe_backend parallel_pandas \" + \\\n",
    "                        \"--calibration_file /reference/env/dev/content/hs37d5/resources/version_control/sherpa/v1.2.3/all_data.BINDING.tab\\n\")   \n",
    "        out_file.write(\"\\n\")\n",
    "            \n",
    "        out_file.write(\"date\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "create_cluster_script_negative(list(input_df.sample_name_tumor))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SAB model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Presentation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['20200825-110238_binding_multi_with_features_on_multi_mono',\n",
       " '20200825-110238_presentation_multi_with_features_on_multi_mono']"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "['20200723-113359_binding_multi_with_features_on_multi_mono',\n",
    "'20200728-032541_presentation_multi_with_features_on_multi_mono']\n",
    "\n",
    "['20200825-110238_binding_multi_with_features_on_multi_mono',\n",
    "'20200825-110238_presentation_multi_with_features_on_multi_mono']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_cluster_script_presentation(patients):\n",
    "    \n",
    "    new_script_file = '/rnd/users/rpyke/code/03-public_MS/scripts/run_sherpa.nextdb.presentation.sh'\n",
    "    \n",
    "    with open(new_script_file, 'w') as out_file:\n",
    "        out_file.write(\"#! /bin/csh\\n\")\n",
    "        out_file.write(\"#$ -S /bin/csh\\n\")\n",
    "        out_file.write(\"#$ -cwd\\n\")\n",
    "        out_file.write(\"#$ -V\\n\")\n",
    "        out_file.write(\"#$ -pe shm 4\\n\")\n",
    "        out_file.write(\"#$ -l h_vmem=10G\\n\")\n",
    "        out_file.write(\"#$ -N SHERPA_inova\\n\")\n",
    "        #out_file.write(\"#$ -t 1-1\\n\".format(len(projects)))\n",
    "        out_file.write(\"#$ -t 1-{0}\\n\".format(len(patients)))\n",
    "        out_file.write(\"#$ -tc 3\\n\")\n",
    "        out_file.write(\"#$ -q all.q\\n\")\n",
    "        out_file.write(\"#$ -o /rnd/users/rpyke/data/03-public_MS/sge_system_files/\\n\")\n",
    "        out_file.write(\"#$ -e /rnd/users/rpyke/data/03-public_MS/sge_system_files\\n\")\n",
    "        out_file.write(\"\\n\")\n",
    "\n",
    "        out_file.write(\"set patients=({0})\\n\".format(\" \".join(patients)))\n",
    "        out_file.write(\"\\n\")\n",
    "    \n",
    "        out_file.write(\"set patient=$patients[$SGE_TASK_ID]\\n\")\n",
    "        out_file.write(\"\\n\")\n",
    "\n",
    "        out_file.write(\"date\\n\")\n",
    "        out_file.write(\"hostname\\n\")\n",
    "        out_file.write(\"echo $patient\")\n",
    "        out_file.write(\"\\n\")\n",
    "    \n",
    "        # SNV\n",
    "        out_file.write(\"cd /rnd/users/mp-wsrv-hmc01/sherpa/runs-M02/20200825-110238_presentation_multi_with_features_on_multi_mono/models/1598386589848858__20200825-110238_presentation_multi_with_features_on_multi_mono__1220ab4ec0834a3bb5a8a15f3ca7d492/deploy_dir/\\n\")\n",
    "        out_file.write(\"python /rnd/users/mp-wsrv-hmc01/sherpa/runs-M02/20200825-110238_presentation_multi_with_features_on_multi_mono/models/1598386589848858__20200825-110238_presentation_multi_with_features_on_multi_mono__1220ab4ec0834a3bb5a8a15f3ca7d492/deploy_dir/1598386589848858__20200825-110238_presentation_multi_with_features_on_multi_mono__1220ab4ec0834a3bb5a8a15f3ca7d492.predict.py \" + \\\n",
    "                        \"--in_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_input/$patient.snv.tab \" + \\\n",
    "                        \"--out_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_output/$patient.snv.PRESENTATION.vMAN.tab \" + \\\n",
    "                        \"--max_mem 30000000000 --max_cpus 4 --dataframe_backend parallel_pandas \" + \\\n",
    "                        \"--calibration_file /rnd/users/mp-wsrv-hmc01/sherpa/rank/20200825-110238_presentation_multi_with_features_on_multi_mono/all_data.20200825-110238_presentation_multi_with_features_on_multi_mono.tab\\n\")   \n",
    "        out_file.write(\"\\n\")\n",
    "                  \n",
    "        # INDEL\n",
    "        out_file.write(\"cd /rnd/users/mp-wsrv-hmc01/sherpa/runs-M02/20200825-110238_binding_multi_with_features_on_multi_mono/models/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6/deploy_dir/\\n\")\n",
    "        out_file.write(\"python /rnd/users/mp-wsrv-hmc01/sherpa/runs-M02/20200825-110238_binding_multi_with_features_on_multi_mono/models/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6/deploy_dir/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6.predict.py \" + \\\n",
    "                        \"--in_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_input/$patient.indel.tab \" + \\\n",
    "                        \"--out_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_output/$patient.indel.PRESENTATION.vMAN.tab \" + \\\n",
    "                        \"--max_mem 30000000000 --max_cpus 4 --dataframe_backend parallel_pandas \" + \\\n",
    "                        \"--calibration_file /rnd/users/mp-wsrv-hmc01/sherpa/rank/20200825-110238_binding_multi_with_features_on_multi_mono/all_data.20200825-110238_binding_multi_with_features_on_multi_mono.tab\\n\")   \n",
    "        out_file.write(\"\\n\")\n",
    "                  \n",
    "        \n",
    "        # FUSION\n",
    "        out_file.write(\"cd /rnd/users/mp-wsrv-hmc01/sherpa/runs-M02/20200825-110238_binding_multi_with_features_on_multi_mono/models/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6/deploy_dir/\\n\")\n",
    "        out_file.write(\"python /rnd/users/mp-wsrv-hmc01/sherpa/runs-M02/20200825-110238_binding_multi_with_features_on_multi_mono/models/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6/deploy_dir/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6.predict.py \" + \\\n",
    "                        \"--in_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_input/$patient.fusion.tab \" + \\\n",
    "                        \"--out_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_output/$patient.fusion.PRESENTATION.vMAN.tab \" + \\\n",
    "                        \"--max_mem 30000000000 --max_cpus 4 --dataframe_backend parallel_pandas \" + \\\n",
    "                        \"--calibration_file /rnd/users/mp-wsrv-hmc01/sherpa/rank/20200825-110238_binding_multi_with_features_on_multi_mono/all_data.20200825-110238_binding_multi_with_features_on_multi_mono.tab\\n\")      \n",
    "        out_file.write(\"\\n\")\n",
    "        \n",
    "        # BASELINE\n",
    "        out_file.write(\"cd /rnd/users/mp-wsrv-hmc01/sherpa/runs-M02/20200825-110238_binding_multi_with_features_on_multi_mono/models/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6/deploy_dir/\\n\")\n",
    "        out_file.write(\"python /rnd/users/mp-wsrv-hmc01/sherpa/runs-M02/20200825-110238_binding_multi_with_features_on_multi_mono/models/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6/deploy_dir/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6.predict.py \" + \\\n",
    "                        \"--in_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_input/$patient.baseline.tab \" + \\\n",
    "                        \"--out_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_output/$patient.baseline.PRESENTATION.vMAN.tab \" + \\\n",
    "                        \"--max_mem 30000000000 --max_cpus 4 --dataframe_backend parallel_pandas \" + \\\n",
    "                        \"--calibration_file /rnd/users/mp-wsrv-hmc01/sherpa/rank/20200825-110238_binding_multi_with_features_on_multi_mono/all_data.20200825-110238_binding_multi_with_features_on_multi_mono.tab\\n\")   \n",
    "        out_file.write(\"\\n\")\n",
    "            \n",
    "        out_file.write(\"date\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "create_cluster_script_presentation(list(input_df.sample_name_tumor))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "BINDING Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_cluster_script_binding(patients):\n",
    "    \n",
    "    new_script_file = '/rnd/users/rpyke/code/03-public_MS/scripts/run_sherpa.nextdb.binding.sh'\n",
    "    \n",
    "    with open(new_script_file, 'w') as out_file:\n",
    "        out_file.write(\"#! /bin/csh\\n\")\n",
    "        out_file.write(\"#$ -S /bin/csh\\n\")\n",
    "        out_file.write(\"#$ -cwd\\n\")\n",
    "        out_file.write(\"#$ -V\\n\")\n",
    "        out_file.write(\"#$ -pe shm 4\\n\")\n",
    "        out_file.write(\"#$ -l h_vmem=10G\\n\")\n",
    "        out_file.write(\"#$ -N SHERPA_inova\\n\")\n",
    "        #out_file.write(\"#$ -t 1-1\\n\".format(len(projects)))\n",
    "        out_file.write(\"#$ -t 1-{0}\\n\".format(len(patients)))\n",
    "        out_file.write(\"#$ -tc 3\\n\")\n",
    "        out_file.write(\"#$ -q all.q\\n\")\n",
    "        out_file.write(\"#$ -o /rnd/users/rpyke/data/03-public_MS/sge_system_files/\\n\")\n",
    "        out_file.write(\"#$ -e /rnd/users/rpyke/data/03-public_MS/sge_system_files\\n\")\n",
    "        out_file.write(\"\\n\")\n",
    "\n",
    "        out_file.write(\"set patients=({0})\\n\".format(\" \".join(patients)))\n",
    "        out_file.write(\"\\n\")\n",
    "    \n",
    "        out_file.write(\"set patient=$patients[$SGE_TASK_ID]\\n\")\n",
    "        out_file.write(\"\\n\")\n",
    "\n",
    "        out_file.write(\"date\\n\")\n",
    "        out_file.write(\"hostname\\n\")\n",
    "        out_file.write(\"echo $patient\")\n",
    "        out_file.write(\"\\n\")\n",
    "    \n",
    "        # SNV\n",
    "        out_file.write(\"cd /rnd/users/mp-wsrv-hmc01/sherpa/runs-M02/20200825-110238_binding_multi_with_features_on_multi_mono/models/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6/deploy_dir/\\n\")\n",
    "        out_file.write(\"python /rnd/users/mp-wsrv-hmc01/sherpa/runs-M02/20200825-110238_binding_multi_with_features_on_multi_mono/models/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6/deploy_dir/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6.predict.py \" + \\\n",
    "                        \"--in_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_input/$patient.snv.tab \" + \\\n",
    "                        \"--out_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_output/$patient.snv.BINDING.vMAN.tab \" + \\\n",
    "                        \"--max_mem 30000000000 --max_cpus 4 --dataframe_backend parallel_pandas \" + \\\n",
    "                        \"--calibration_file /rnd/users/mp-wsrv-hmc01/sherpa/rank/20200825-110238_binding_multi_with_features_on_multi_mono/all_data.20200825-110238_binding_multi_with_features_on_multi_mono.tab\\n\")   \n",
    "        out_file.write(\"\\n\")\n",
    "                  \n",
    "        # INDEL\n",
    "        out_file.write(\"cd /rnd/users/mp-wsrv-hmc01/sherpa/runs-M02/20200825-110238_binding_multi_with_features_on_multi_mono/models/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6/deploy_dir/\\n\")\n",
    "        out_file.write(\"python /rnd/users/mp-wsrv-hmc01/sherpa/runs-M02/20200825-110238_binding_multi_with_features_on_multi_mono/models/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6/deploy_dir/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6.predict.py \" + \\\n",
    "                        \"--in_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_input/$patient.indel.tab \" + \\\n",
    "                        \"--out_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_output/$patient.indel.BINDING.vMAN.tab \" + \\\n",
    "                        \"--max_mem 30000000000 --max_cpus 4 --dataframe_backend parallel_pandas \" + \\\n",
    "                        \"--calibration_file /rnd/users/mp-wsrv-hmc01/sherpa/rank/20200825-110238_binding_multi_with_features_on_multi_mono/all_data.20200825-110238_binding_multi_with_features_on_multi_mono.tab\\n\")   \n",
    "        out_file.write(\"\\n\")\n",
    "                  \n",
    "        \n",
    "        # FUSION\n",
    "        out_file.write(\"cd /rnd/users/mp-wsrv-hmc01/sherpa/runs-M02/20200825-110238_binding_multi_with_features_on_multi_mono/models/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6/deploy_dir/\\n\")\n",
    "        out_file.write(\"python /rnd/users/mp-wsrv-hmc01/sherpa/runs-M02/20200825-110238_binding_multi_with_features_on_multi_mono/models/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6/deploy_dir/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6.predict.py \" + \\\n",
    "                        \"--in_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_input/$patient.fusion.tab \" + \\\n",
    "                        \"--out_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_output/$patient.fusion.BINDING.vMAN.tab \" + \\\n",
    "                        \"--max_mem 30000000000 --max_cpus 4 --dataframe_backend parallel_pandas \" + \\\n",
    "                        \"--calibration_file /rnd/users/mp-wsrv-hmc01/sherpa/rank/20200825-110238_binding_multi_with_features_on_multi_mono/all_data.20200825-110238_binding_multi_with_features_on_multi_mono.tab\\n\")      \n",
    "        out_file.write(\"\\n\")\n",
    "        \n",
    "        # BASELINE\n",
    "        out_file.write(\"cd /rnd/users/mp-wsrv-hmc01/sherpa/runs-M02/20200825-110238_binding_multi_with_features_on_multi_mono/models/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6/deploy_dir/\\n\")\n",
    "        out_file.write(\"python /rnd/users/mp-wsrv-hmc01/sherpa/runs-M02/20200825-110238_binding_multi_with_features_on_multi_mono/models/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6/deploy_dir/15983865633686793__20200825-110238_binding_multi_with_features_on_multi_mono__9430ba70afe34e08b0066d673c3478e6.predict.py \" + \\\n",
    "                        \"--in_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_input/$patient.baseline.tab \" + \\\n",
    "                        \"--out_file /rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_output/$patient.baseline.BINDING.vMAN.tab \" + \\\n",
    "                        \"--max_mem 30000000000 --max_cpus 4 --dataframe_backend parallel_pandas \" + \\\n",
    "                        \"--calibration_file /rnd/users/mp-wsrv-hmc01/sherpa/rank/20200825-110238_binding_multi_with_features_on_multi_mono/all_data.20200825-110238_binding_multi_with_features_on_multi_mono.tab\\n\")   \n",
    "        out_file.write(\"\\n\")\n",
    "            \n",
    "        out_file.write(\"date\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "create_cluster_script_binding(list(input_df.sample_name_tumor))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Concating all epitopes into a single file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Getting all mutation types into a single file for each patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BINDING\n",
      "0 40425-0012_Pre_Tumor_DNA\n",
      "1 40425-0008_Pre_Tumor_DNA\n",
      "2 40425-0014_Post_Tumor_DNA\n",
      "3 40425-0014_Pre_Tumor_DNA\n",
      "4 40425-0001_Post_Tumor_DNA\n",
      "5 40425-0001_Pre_TumorDNA_ACE3\n",
      "6 40425-0002_Post_Tumor_DNA\n",
      "7 40425-0011_Pre_Tumor_DNA\n",
      "8 40425-0009_Post_Tumor_DNA\n",
      "9 40425-0009_Pre_Tumor_DNA\n",
      "10 40425-0011_Post_Tumor_DNA\n",
      "11 7140_Post_Tumor_DNA\n",
      "12 40425-0013_Pre_Tumor_DNA\n",
      "13 40425-0013_Post_Tumor_DNA\n",
      "14 40425-0008_Post_Tumor_DNA\n",
      "15 40425-0002_Pre_TumorDNA_ACE3\n",
      "PRESENTATION\n",
      "0 40425-0012_Pre_Tumor_DNA\n",
      "1 40425-0008_Pre_Tumor_DNA\n",
      "2 40425-0014_Post_Tumor_DNA\n",
      "3 40425-0014_Pre_Tumor_DNA\n",
      "4 40425-0001_Post_Tumor_DNA\n",
      "5 40425-0001_Pre_TumorDNA_ACE3\n",
      "6 40425-0002_Post_Tumor_DNA\n",
      "7 40425-0011_Pre_Tumor_DNA\n",
      "8 40425-0009_Post_Tumor_DNA\n",
      "9 40425-0009_Pre_Tumor_DNA\n",
      "10 40425-0011_Post_Tumor_DNA\n",
      "11 7140_Post_Tumor_DNA\n",
      "12 40425-0013_Pre_Tumor_DNA\n",
      "13 40425-0013_Post_Tumor_DNA\n",
      "14 40425-0008_Post_Tumor_DNA\n",
      "15 40425-0002_Pre_TumorDNA_ACE3\n"
     ]
    }
   ],
   "source": [
    "for model in ['BINDING', 'PRESENTATION']:\n",
    "    print(model)\n",
    "    for i, sample in enumerate(list(input_df.sample_name_tumor)):\n",
    "        print(i, sample)\n",
    "        files = [x for x in os.listdir('/rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_output') if (sample in x) and ('vMAN' in x)]\n",
    "        if len(files) > 0:\n",
    "            all_tmps = []\n",
    "            for f in [k for k in files if model in k]:\n",
    "                tmp_df = pd.read_csv('/rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/sherpa_output/{0}'.format(f), sep='\\t')\n",
    "                all_tmps.append(tmp_df)\n",
    "            all_df = pd.concat(all_tmps, sort=True)\n",
    "\n",
    "            all_df.to_csv('/rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/merged_output/{0}.{1}.all.vMAN.tab'.format(sample, model), sep='\\t', index=False)\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assess which mutations are new"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = 'PRESENTATION'\n",
    "sample_pre = '40425-0013_Pre_Tumor_DNA'\n",
    "sample_post = '40425-0013_Post_Tumor_DNA'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pre = pd.read_csv('/rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/merged_output/{0}.{1}.all.tab'.format(sample_pre, model),\n",
    "                sep='\\t')\n",
    "df_post = pd.read_csv('/rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/merged_output/{0}.{1}.all.tab'.format(sample_post, model),\n",
    "                sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_novel = df_post[~df_post.pepseq.isin(list(df_pre.pepseq))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "HLA-B55:01    1\n",
       "HLA-B44:03    1\n",
       "HLA-C07:06    1\n",
       "Name: allele, dtype: int64"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_novel[df_novel.sherpa_calibrated < 0.5].allele.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Expanding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PRESENTATION\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'results' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-31-ef202b8d530d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmodel\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mall_dfs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample\u001b[0m \u001b[0;32min\u001b[0m \u001b[0menumerate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mresults\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'sample'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mi\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0;36m10\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m         \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'results' is not defined"
     ]
    }
   ],
   "source": [
    "model = 'PRESENTATION'\n",
    "\n",
    "print(model)\n",
    "all_dfs = []\n",
    "for i, sample in enumerate(results['sample'].unique()):\n",
    "    if i % 10 == 0:\n",
    "        print(i, sample)\n",
    "    try:\n",
    "        tmp_df = pd.read_csv('/rnd/users/rpyke/data/03-public_MS/applications/NeXTDB/merged_output/{0}.{1}.all.tab'.format(sample, model), sep='\\t')\n",
    "        tmp_df['sample'] = sample.split('_')[0]\n",
    "        tmp_df['Allele'] = tmp_df.allele\n",
    "        tmp_df['Rank'] = tmp_df.sherpa_calibrated\n",
    "        tmp_df['Peptide'] = tmp_df.pepseq\n",
    "        tmp_df['time'] = sample.split('_')[1]\n",
    "        all_dfs.append(tmp_df)\n",
    "    except:\n",
    "        print(sample)\n",
    "\n",
    "neo_df = pd.concat(all_dfs)\n",
    "\n",
    "# For neoantigens\n",
    "#neo_df = neo_df.sort_values('Rank').drop_duplicates(['#peptide_id', 'Allele'])\n",
    "\n",
    "neo_df_restricted = neo_df[['Peptide', 'Rank', 'Allele', 'sample', 'time']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
